### PR TITLE
Custom multiclass objectives in R

### PR DESCRIPTION
**Pull request by Someone894**
Added workaround described in https://github.com/dmlc/xgboost/issues/2776

with this change the https://github.com/pmontman/fforma package works fine.

### CODE REVIEW

**Comment by trivialfis** [comment]
Could you please help taking a look @pommedeterresautee ?

**Comment by pommedeterresautee** [comment]
Tks, I check today or tomorrow.

**Comment by pommedeterresautee** [comment]
@Someone894 Hi, I have made few checks and it seems to not break simple cases (I was worried about the reshaping).
However, I am wondering how `outputmargin` set to TRUE will never impact in some way the training? (in some other cases that your own custom objective).
I am asking because the function `xgb.iter.update` is obviously central and I am under the impression you have already performed a deep work for `fforma`, so may be you can help me in better understand that point.

**Comment by Someone894** [comment]
The `FFORMA`-System needs the custom objective in order to work. So @pmontman created a custom-version (https://github.com/pmontman/customxgboost), which is not maintained anymore. This got me worried about the old code causing some trouble in the future. So I talked to him in order to understand what his changes to the basic XGBoost-package where. Then I implemented these changes and tested the `FFORMA`-System with them and it worked fine. but to be honest I did not understand the changes in the  XGBoost-package since it feels really complicated  to me.

**Comment by pommedeterresautee** [comment]
Thank you a lot for these information.
Basically the function `xgb.iter.update` is quite important.
It is called at each iteration of the boosting process.
https://github.com/dmlc/xgboost/blob/5e97de6a4186cd616bda3387c8cf9c5ec1904189/R-package/R/xgb.train.R#L331
Its purpose is to apply the objective to the predictions of the model being built and compute the loss.
https://github.com/dmlc/xgboost/blob/9536a7e0d216bc3f0487b91ede4597acc391272c/R-package/R/utils.R#L149
This PR modifies the predictions by setting `outputmargin` to `True`, basically the predictions are now the margin directly computed by XGBoost (the original untransformed predictions).
My feeling is that in several cases this will have unwanted effect on the objective, and the quality of the model.

Would it be possible for you to contact @pmontman to see if there are reasons why such change would be Ok?

Otherwise, I am thinking that another approach may be more appropriate. 
May be another hook function or some information provided through parameters of `xgb.iter.update`  making possible for third party objectives to retrieve the margin.

Regarding the reshape thing, it should be easy to perform in the objective itself.

**Comment by jeffzi** [comment]
I encountered the same issue several months ago while looking at FFORMA as well (see #4288).

I had to make the same modification in xgb.iter.eval to fix xgb.cv as well.
https://github.com/dmlc/xgboost/blob/e089e16e3da6637c5374eb1e2e0a3b9fd92907ba/R-package/R/utils.R#L175

**Comment by pommedeterresautee** [comment]
Thanks @jeffzi . I understand the purpose of the change, my question is how this change may not have any impact on other use cases? because here we are wondering if it makes sense to apply the change to the main library.
I may miss something.

**Comment by Someone894** [comment]
I tried to reach @pmontman to also have a look at this topic, but nothing happened, as you can see.

**Comment by trivialfis** [comment]
@Someone894 @pommedeterresautee It make sense to me to use raw prediction as objective function should handle this transformation itself.  WDYT?

**Comment by jeffzi** [comment]
It makes sense to me to use raw prediction, since it reflects what happens when no custom objective is given.

https://github.com/dmlc/xgboost/blob/ad4a1c732cfad6d6dc5be012145a2dac37413cc7/R-package/R/utils.R#L145-L151

`UpdateOneIter`  uses raw prediction:

https://github.com/dmlc/xgboost/blob/ad4a1c732cfad6d6dc5be012145a2dac37413cc7/src/learner.cc#L561-L567

**Comment by seangupta** [comment]
Hi, I've encountered this problem as well. Will there be an update soon? Thanks.

**Comment by trivialfis** [comment]
I incline to merge this PR(or new one with tests).  @hcho3 WDYT?
